loadtest/loadtest.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) 2019-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package loadtest
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/mattermost/mattermost-load-test-ng/config"
+	"github.com/mattermost/mattermost-load-test-ng/loadtest/control"
+
+	"github.com/mattermost/mattermost-server/v5/mlog"
+)
+
+// LoadTester is a structure holding all the state needed to run a load-test
+type LoadTester struct {
+	controllers   []control.UserController
+	config        *config.LoadTestConfig
+	wg            sync.WaitGroup
+	status        chan control.UserStatus
+	started       bool
+	newController NewController
+}
+
+// NewController is a factory function that returns a new
+// control.UserController given an id and a channel of control.UserStatus
+type NewController func(int, chan<- control.UserStatus) control.UserController
+
+func (lt *LoadTester) handleStatus() {
+	for us := range lt.status {
+		if us.Code == control.USER_STATUS_STOPPED || us.Code == control.USER_STATUS_FAILED {
+			lt.wg.Done()
+		}
+		if us.Code == control.USER_STATUS_ERROR {
+			mlog.Info(us.Err.Error(), mlog.Int("controller_id", us.ControllerId))
+			continue
+		} else if us.Code == control.USER_STATUS_FAILED {
+			mlog.Error(us.Err.Error())
+			continue
+		}
+		mlog.Info(us.Info, mlog.Int("controller_id", us.ControllerId))
+	}
+}
+
+// AddUser increments by one the number of concurrently active users
+func (lt *LoadTester) AddUser() error {
```

Comment:
Should we also add wrapper functions which add/remove n users ?

Reply:
I've intentionally kept it as simple as possible in order to let the user of the API be in control of the timings related to adding/removing users.  I believe it should be the higher level to decide how to add/remove users (e.g. all at once vs n per second or some other random way).